File: src/flowsheet/Flowsheet.cpp
#include "chemsim/flowsheet/Flowsheet.hpp"
#include "chemsim/io/FlowsheetParser.hpp"
#include <fstream>
#include <iomanip>
#include <numeric>
#include <sstream>
#include <stdexcept>

namespace chemsim {

namespace {

std::string phaseToString(Phase phase) {
    switch (phase) {
    case Phase::VAPOR: return "VAPOR";
    case Phase::LIQUID: return "LIQUID";
    case Phase::MIXED: return "MIXED";
    case Phase::UNKNOWN: return "UNKNOWN";
    }
    return "UNKNOWN";
}

nlohmann::json streamToJson(const Stream& stream) {
    return {
        {"name", stream.name},
        {"T", stream.T},
        {"P", stream.P},
        {"totalFlow", stream.totalFlow},
        {"z", stream.z},
        {"phase", phaseToString(stream.phase)},
        {"vaporFraction", stream.vaporFraction},
        {"x", stream.x},
        {"y", stream.y},
        {"H", stream.H},
        {"S", stream.S}
    };
}

} // namespace

Flowsheet::Flowsheet(std::vector<Component> components)
    : components_(std::move(components)) {
    if (components_.empty())
        throw std::invalid_argument("Flowsheet: components must not be empty");
    eos_ = std::make_unique<PengRobinson>(components_);
    flash_ = std::make_unique<FlashCalculator>(*eos_, components_);
}

std::vector<double> Flowsheet::compositionVector(
    const std::map<std::string, double>& composition) const {
    std::vector<double> z(components_.size(), 0.0);
    double sum = 0.0;

    for (std::size_t i = 0; i < components_.size(); ++i) {
        const auto it = composition.find(components_[i].id);
        if (it == composition.end())
            throw std::invalid_argument(
                "Flowsheet: missing composition for component '" + components_[i].id + "'");
        z[i] = it->second;
        sum += z[i];
    }

    if (sum <= 0.0)
        throw std::invalid_argument("Flowsheet: composition sum must be > 0");

    for (double& zi : z) zi /= sum;
    return z;
}

void Flowsheet::initializeThermo(Stream& stream) const {
    if (stream.totalFlow <= 0.0 || stream.z.empty()) return;

    auto flash_result = flash_->flashTP(stream.T, stream.P, stream.z);
    stream.vaporFraction = flash_result.beta;
    stream.x = flash_result.x;
    stream.y = flash_result.y;
    stream.phase = (flash_result.beta < 1e-10) ? Phase::LIQUID
                 : (flash_result.beta > 1.0 - 1e-10) ? Phase::VAPOR
                 : Phase::MIXED;
    stream.H = flash_->totalEnthalpy(flash_result);
    stream.S = flash_->totalEntropy(flash_result);
}

Stream& Flowsheet::addStream(const std::string& name,
                             double T, double P, double flow,
                             const std::map<std::string, double>& composition) {
    Stream stream;
    stream.name = name;
    stream.T = T;
    stream.P = P;
    stream.totalFlow = flow;
    stream.z = compositionVector(composition);
    initializeThermo(stream);

    streams_[name] = stream;
    graph_.setFeed(name, stream);
    return streams_.at(name);
}

void Flowsheet::addUnit(const std::string& name, std::unique_ptr<IUnitOp> op) {
    if (name.empty())
        throw std::invalid_argument("Flowsheet: unit name must not be empty");
    graph_.addUnit(name, std::move(op));
    unit_names_.push_back(name);
}

void Flowsheet::ensureStreamExists(const std::string& name) {
    if (name.empty()) return;
    if (!streams_.count(name)) {
        Stream s;
        s.name = name;
        streams_[name] = s;
    }
}

void Flowsheet::connect(const std::string& stream_name,
                        const std::string& from_unit, const std::string& from_port,
                        const std::string& to_unit,   const std::string& to_port) {
    ensureStreamExists(stream_name);
    graph_.connect(stream_name, from_unit, from_port, to_unit, to_port);
}

bool Flowsheet::solve() {
    return solve(RecycleSolver::Options{});
}

bool Flowsheet::solve(RecycleSolver::Options opts) {
    RecycleSolver solver(graph_, opts);
    auto result = solver.solve(streams_);
    streams_ = std::move(result.streams);
    return result.converged;
}

const Stream& Flowsheet::getStream(const std::string& name) const {
    return streams_.at(name);
}

Stream& Flowsheet::getStream(const std::string& name) {
    return streams_.at(name);
}

IUnitOp& Flowsheet::getUnit(const std::string& name) {
    return graph_.unit(name);
}

const IUnitOp& Flowsheet::getUnit(const std::string& name) const {
    return graph_.unit(name);
}

std::vector<std::string> Flowsheet::streamNames() const {
    std::vector<std::string> names;
    names.reserve(streams_.size());
    for (const auto& [name, _] : streams_)
        names.push_back(name);
    return names;
}

std::string Flowsheet::summary() const {
    std::ostringstream out;
    out << std::fixed << std::setprecision(3);
    out << "ChemSim Flowsheet Summary\n";
    out << "Components:";
    for (const auto& component : components_)
        out << ' ' << component.id;
    out << "\nStreams (" << streams_.size() << ")\n";
    for (const auto& [name, stream] : streams_) {
        out << "  " << name
            << ": F=" << stream.totalFlow << " mol/s"
            << ", T=" << stream.T << " K"
            << ", P=" << stream.P << " Pa"
            << ", phase=" << phaseToString(stream.phase)
            << ", beta=" << stream.vaporFraction
            << '\n';
    }
    return out.str();
}

void Flowsheet::printSummary(std::ostream& os) const {
    os << summary();
}

nlohmann::json Flowsheet::resultsAsJson() const {
    nlohmann::json result;
    result["components"] = nlohmann::json::array();
    for (const auto& component : components_) {
        result["components"].push_back({
            {"id", component.id},
            {"name", component.name},
            {"MW", component.MW},
            {"Tc", component.Tc},
            {"Pc", component.Pc},
            {"omega", component.omega}
        });
    }

    result["streams"] = nlohmann::json::object();
    for (const auto& [name, stream] : streams_)
        result["streams"][name] = streamToJson(stream);
    return result;
}

void Flowsheet::exportResults(const std::string& json_path) const {
    std::ofstream out(json_path);
    if (!out.is_open())
        throw std::runtime_error("Flowsheet: cannot open results path " + json_path);
    out << resultsAsJson().dump(2);
}

Flowsheet Flowsheet::fromJSON(const std::string& json_path,
                              const std::string& component_db_path) {
    return FlowsheetParser::parseFile(json_path, component_db_path);
}

std::unique_ptr<Flowsheet> Flowsheet::fromJSONUnique(const std::string& json_path,
                                                     const std::string& component_db_path) {
    return FlowsheetParser::parseFileUnique(json_path, component_db_path);
}

} // namespace chemsim

File: chemsim/flowsheet/flowsheet.py
"""Top-level Flowsheet orchestrator."""
from __future__ import annotations

import copy
import json
from typing import Dict, List, Optional

from chemsim.core import Component, Phase, Stream
from chemsim.flowsheet.graph import FlowsheetGraph
from chemsim.flowsheet.recycle import RecycleOptions, RecycleSolver
from chemsim.ops.base import IUnitOp
from chemsim.ops.distillation import DistillationColumnOp
from chemsim.thermo.flash import FlashCalculator
from chemsim.thermo.peng_robinson import PengRobinson


class Flowsheet:
    def __init__(self, components: List[Component]):
        if not components:
            raise ValueError("Flowsheet: components must not be empty")
        self._components = list(components)
        self._eos = PengRobinson(self._components)
        self._flash = FlashCalculator(self._eos, self._components)
        self._graph = FlowsheetGraph()
        self._streams: Dict[str, Stream] = {}
        self._base_streams: Dict[str, Stream] = {}

    # ── Accessors ─────────────────────────────────────────────────────────────

    def components(self) -> List[Component]:
        return list(self._components)

    def flash_calculator(self) -> FlashCalculator:
        return self._flash

    # ── Build ─────────────────────────────────────────────────────────────────

    def add_stream(self, name: str, T: float, P: float, flow: float,
                   composition: Dict[str, float]) -> Stream:
        z = self._composition_vector(composition)
        s = Stream(name=name, T=T, P=P, total_flow=flow, z=z)
        self._initialize_thermo(s)
        self._streams[name] = s
        self._base_streams[name] = copy.deepcopy(s)
        self._graph.set_feed(name, s)
        return self._streams[name]

    def add_unit(self, name: str, op: IUnitOp) -> None:
        if not name:
            raise ValueError("Flowsheet: unit name must not be empty")
        self._graph.add_unit(name, op)

    def connect(self, stream_name: str,
                from_unit: str = "", from_port: str = "",
                to_unit: str = "", to_port: str = "") -> None:
        if stream_name and stream_name not in self._streams:
            s = Stream(name=stream_name)
            self._streams[stream_name] = s
        self._graph.connect(stream_name, from_unit, from_port, to_unit, to_port)

    # ── Solve ─────────────────────────────────────────────────────────────────

    def solve(self, opts: Optional[RecycleOptions] = None) -> bool:
        solver = RecycleSolver(self._graph, opts or RecycleOptions())
        result = solver.solve(self._streams)
        self._streams = result.streams
        return result.converged

    # ── Stream / unit access ──────────────────────────────────────────────────

    def get_stream(self, name: str) -> Stream:
        return self._streams[name]

    def stream_names(self) -> List[str]:
        return list(self._streams.keys())

    def get_unit(self, name: str) -> IUnitOp:
        return self._graph.unit(name)

    # ── RL interface ──────────────────────────────────────────────────────────

    def set_param(self, unit_name: str, param_name: str, value: float) -> None:
        op = self._graph.unit(unit_name)
        if isinstance(op, DistillationColumnOp):
            if param_name == "refluxRatio":
                op.set_reflux_ratio(value); return
            if param_name == "distillateFrac":
                op.set_distillate_frac(value); return
        raise ValueError(
            f"Flowsheet.set_param: unknown unit '{unit_name}' "
            f"or param '{param_name}'")

    def set_stream_conditions(self, stream_name: str, T: float, P: float) -> None:
        s = self._streams[stream_name]
        s.T = T; s.P = P
        self._initialize_thermo(s)
        self._graph.set_feed(stream_name, s)

    def reset_to_base(self) -> None:
        self._streams = copy.deepcopy(self._base_streams)
        for name, s in self._base_streams.items():
            self._graph.set_feed(name, copy.deepcopy(s))

    def get_unit_scalar(self, unit_name: str, key: str) -> float:
        op = self._graph.unit(unit_name)
        if isinstance(op, DistillationColumnOp):
            return op.get_scalar(key)
        raise ValueError(
            f"Flowsheet.get_unit_scalar: unknown unit '{unit_name}' or key '{key}'")

    # ── Output ────────────────────────────────────────────────────────────────

    def summary(self) -> str:
        lines = ["ChemSim Flowsheet Summary"]
        lines.append("Components: " + " ".join(c.id for c in self._components))
        lines.append(f"Streams ({len(self._streams)})")
        for name, s in self._streams.items():
            lines.append(
                f"  {name}: F={s.total_flow:.3f} mol/s"
                f", T={s.T:.3f} K"
                f", P={s.P:.3f} Pa"
                f", phase={s.phase.value}"
                f", beta={s.vapor_fraction:.3f}")
        return "\n".join(lines)

    def results_as_json(self) -> str:
        data = {
            "components": [
                {"id": c.id, "name": c.name, "MW": c.MW,
                 "Tc": c.Tc, "Pc": c.Pc, "omega": c.omega}
                for c in self._components
            ],
            "streams": {name: s.to_dict()
                        for name, s in self._streams.items()},
        }
        return json.dumps(data, indent=2)

    def export_results(self, json_path: str) -> None:
        with open(json_path, "w") as f:
            f.write(self.results_as_json())

    def results_table(self) -> List[dict]:
        return [s.to_dict() for s in self._streams.values()]

    def __repr__(self) -> str:
        names = list(self._streams.keys())
        return f"<Flowsheet streams={names}>"

    # ── Factory ───────────────────────────────────────────────────────────────

    @staticmethod
    def create(component_ids: List[str], db_path: str) -> Flowsheet:
        from chemsim.io.component_db import ComponentDB
        db = ComponentDB(db_path)
        return Flowsheet(db.get(component_ids))

    @staticmethod
    def from_json(json_path: str, component_db_path: str) -> Flowsheet:
        from chemsim.io.parser import FlowsheetParser
        return FlowsheetParser.parse_file(json_path, component_db_path)

    # ── Internal helpers ──────────────────────────────────────────────────────

    def _composition_vector(self, composition: Dict[str, float]) -> List[float]:
        z = []
        total = 0.0
        for c in self._components:
            val = composition.get(c.id)
            if val is None:
                raise ValueError(
                    f"Flowsheet: missing composition for component '{c.id}'")
            z.append(val)
            total += val
        if total <= 0.0:
            raise ValueError("Flowsheet: composition sum must be > 0")
        return [zi / total for zi in z]

    def _initialize_thermo(self, s: Stream) -> None:
        if s.total_flow <= 0.0 or not s.z:
            return
        r = self._flash.flash_TP(s.T, s.P, s.z)
        s.vapor_fraction = r.beta
        s.x = r.x; s.y = r.y
        s.phase = (Phase.LIQUID if r.beta < 1e-10
                   else Phase.VAPOR if r.beta > 1.0 - 1e-10
                   else Phase.MIXED)
        s.H = self._flash.total_enthalpy(r)
        s.S = self._flash.total_entropy(r)
